### PR TITLE
user_path methods take required_fields keyword argument

### PR DESCRIPTION
**Pull request by lavaturtle**
This updates the `user_path` methods for all CRMs to take an optional `required_fields` keyword argument. Support for the deprecated `phone_required` argument is removed. For krautbuster, `petition_id` becomes a keyword argument.

Because this is a breaking change, the major version is increased.

### CODE REVIEW

File: lib/tokyo_api/actionkit.rb [hunk 1]
```diff
@@ -8,8 +8,12 @@ def full_user(id)
       client.get_request("#{normalized_base_path}full_user/#{id}").body
     end
 
-    def user_path(id, phone_required)
-      "/#{normalized_base_path}user/#{id}?phone_required=#{phone_required}"
+    def user_path(id, required_fields: nil)
+      path = "/#{normalized_base_path}user/#{id}"
+      unless required_fields.nil?
+        path << "?required_fields=#{required_fields.join(',')}"
```
Comment (anero) [inline lib/tokyo_api/actionkit.rb]: Do we need to encode the fields to avoid getting an invalid URL?

File: lib/tokyo_api/krautbuster.rb [hunk 2]
```diff
@@ -25,12 +25,16 @@ def subscription_status(token)
...
     end
 
-    def user_path(session_id, petition_id)
-      "/#{normalized_base_path}user/#{session_id}?petition_id=#{petition_id}"
+    def user_path(session_id, petition_id:, required_fields: nil)
+      path = "/#{normalized_base_path}user/#{session_id}?petition_id=#{petition_id}"
+      unless required_fields.nil?
+        path << "&required_fields=#{required_fields.join(',')}"
+      end
+      path
```
Comment (anero) [inline lib/tokyo_api/krautbuster.rb]: We should create a Trello card for Campact to know how to deploy these changes, it should include the commit SHA of agra & tokyo to get the changes consistently deployed.

**Review by anero — APPROVED** [review]
👍

File: lib/tokyo_api/actionkit.rb [hunk 3]
```diff
@@ -8,8 +8,12 @@ def full_user(id)
       client.get_request("#{normalized_base_path}full_user/#{id}").body
     end
 
-    def user_path(id, phone_required)
-      "/#{normalized_base_path}user/#{id}?phone_required=#{phone_required}"
+    def user_path(id, required_fields: nil)
```
Comment (woodhull) [inline lib/tokyo_api/actionkit.rb:11]: not a new problem but is there a remote url retrieval bug here: should we validate the id so someone is not able to insert tricky IDs that are in turn able to trigger weird HTTP GETs behind our firewall?

Comment (woodhull) [reply lib/tokyo_api/actionkit.rb:11]: It probably is not an actually exploitable bug, but maybe worth running the id through a regex to make sure it doesn't have a `://`

File: lib/tokyo_api/actionkit.rb (hunk 1 above)
Comment (lavaturtle) [reply lib/tokyo_api/actionkit.rb]: Good question! I _expect_ this to be called only with valid fields that are made up of lowercase letters and underscores. But it wouldn't hurt to pass them through `URI.escape`

**Comment by woodhull** [comment]
Let's start URI escaping everything untrusted.

File: lib/tokyo_api/krautbuster.rb (hunk 2 above)
Comment (lavaturtle) [reply lib/tokyo_api/krautbuster.rb]: Good point. Is there a particular column this card should go in?
I think they'll mainly need to deploy the updated tokyo before or at the same time as they deploy the updated agra.

File: lib/tokyo_api/actionkit.rb (hunk 3 above)
Comment (lavaturtle) [reply lib/tokyo_api/actionkit.rb:11]: Will address this in https://www.pivotaltracker.com/n/projects/1037834/stories/150811559

File: lib/tokyo_api/krautbuster.rb (hunk 2 above)
Comment (lavaturtle) [reply lib/tokyo_api/krautbuster.rb]: https://trello.com/c/v22wOhEA

**Pull request merged**